tool/tsh/tsh_test.go

```diff
@@ -410,4 +480,111 @@ func TestFormatConnectCommand(t *testing.T) {
 			require.Equal(t, test.command, formatConnectCommand(cluster, test.db))
 		})
 	}
+
+}
+
+func makeTestServers(t *testing.T, bootstrap ...services.Resource) (auth *service.TeleportProcess, proxy *service.TeleportProcess, done func()) {
+	var deferred []func()
+	var success bool
+	cleanup := func() {
+		for _, fn := range deferred {
+			fn()
+		}
+	}
+	defer func() {
+		if !success {
+			cleanup()
+		}
+	}()
+
+	authDir, err := ioutil.TempDir("", "tsh-auth")
+	require.NoError(t, err)
+
+	proxyDir, err := ioutil.TempDir("", "tsh-auth")
+	require.NoError(t, err)
+
+	deferred = append(deferred, func() {
+		os.RemoveAll(authDir)
+		os.RemoveAll(proxyDir)
+	})
+
+	// Set up a test auth server.
+	//
+	// We need this to get a random port assigned to it and allow parallel
+	// execution of this test.
+	cfg := service.MakeDefaultConfig()
+	cfg.Hostname = "localhost"
+	cfg.DataDir = authDir
+
+	cfg.AuthServers = []utils.NetAddr{randomLocalAddr}
+	cfg.Auth.Resources = bootstrap
+	cfg.Auth.StorageConfig.Params = backend.Params{defaults.BackendPath: filepath.Join(cfg.DataDir, defaults.BackendDir)}
+	cfg.Auth.StaticTokens, err = services.NewStaticTokens(services.StaticTokensSpecV2{
+		StaticTokens: []services.ProvisionTokenV1{{
+			Roles:   []teleport.Role{teleport.RoleProxy},
+			Expires: time.Now().Add(time.Minute),
+			Token:   staticToken,
+		}},
+	})
+	require.NoError(t, err)
+	cfg.SSH.Enabled = false
+	cfg.Auth.Enabled = true
+	cfg.Auth.SSHAddr = randomLocalAddr
+	cfg.Proxy.Enabled = false
+
+	auth, err = service.NewTeleport(cfg)
+	require.NoError(t, err)
+	require.NoError(t, auth.Start())
+
+	deferred = append(deferred, func() {
+		auth.Close()
+	})
+
+	// Wait for proxy to become ready.
+	eventCh := make(chan service.Event, 1)
+	auth.WaitForEvent(auth.ExitContext(), service.AuthTLSReady, eventCh)
+	select {
+	case <-eventCh:
+	case <-time.After(10 * time.Second):
+		t.Fatal("auth server didn't start after 10s")
```

Comment:
I've seen similar events tested on multiple occasions (in integration tests) and I don't believe this should be exercised here at all. If you needed to test whether a particular teleport service is able to start - there should be a dedicated test just for that. Here, we don't really care if fails to start after 10s or at all (and triggers the testing to fail after a timeout), we need know when it has started - for which was can simply block on the events chan.
For example, in my case the test fails:
```
...
WARN [PROC:4]    Teleport process has exited with error: teleport process has shutdown service:register.proxy service/supervisor.go:251
--- FAIL: TestMakeClient (14.40s)
    tsh_test.go:583: proxy web server didn't start after 10s
FAIL
FAIL	github.com/gravitational/teleport/tool/tsh	29.510s
?   	github.com/gravitational/teleport/tool/tsh/common	[no test files]
```
which does not mean that this particular test would have failed (which is confusing). Looking further, I stumble on this:
```
ERRO [PROC:2]    Proxy failed to establish connection to cluster: lstat /var/lib/teleport/ca.cert: permission denied, Post "https://127.0.0.1:43901/v1/webapi/host/credentials": x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs. service/connect.go:63
```
which explains the reason for the failure but again has nothing to do with the test at hand.

One might argue that this would be more confusing since it is not clear whether the test is stuck or is still beint built for instance. In that case, I'd encourage to always run the tests with an explicitly set timeout:
```
$ go test -race -count=1 -timeout=20s ./tool/tsh/...

...
ERRO [PROC:2]    Proxy failed to establish connection to cluster: lstat /var/lib/teleport/ca.cert: permission denied, Post "https://127.0.0.1:33731/v1/webapi/host/credentials": x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs. service/connect.go:63
panic: test timed out after 20s

goroutine 493 [running]:
testing.(*M).startAlarm.func1()
	/home/deemok/dev/go/go/src/testing/testing.go:1618 +0x11f
created by time.goFunc
	/home/deemok/dev/go/go/src/time/sleep.go:167 +0x52
```
The benefit in this case is the actual failure is much closer to the stack dump and is much easier to spot.

Reply:
I don't like the idea of having *no* timeout (prefer to not rely on extra flags getting passed to `go test` in order to have tests behave in a sane manner), but I see your point about this being outside of the domain of the tests, and I agree that the stack traces from `-timeout` are helpful.  I bumped the timeout to 30 seconds, and added a comment explaining that reasonable startup times are the domain of other tests.  This will preserve the ability to use `-timeout=20s` to get a nice traceback without permitting the test to lock up for *too* long.